engine/engine_test.go

```diff
@@ -347,6 +354,263 @@ func TestEngine_NotifyPutThenNotifyRemove(t *testing.T) {
+
+	subject, err := engine.New()
+	require.NoError(t, err)
+	err = subject.Start(ctx)
+	require.NoError(t, err)
+	defer subject.Shutdown()
+
+	wantContextID := []byte("fish")
+	subject.RegisterMultihashLister(func(ctx context.Context, p peer.ID, contextID []byte) (provider.MultihashIterator, error) {
+		if string(contextID) == string(wantContextID) {
+			return provider.SliceMultihashIterator(mhs), nil
+		}
+		return nil, errors.New("not found")
+	})
+
+	gotPutAdCid1, err := subject.NotifyPut(ctx, defaultProvider, wantContextID, metadata.New(metadata.Bitswap{}))
+	require.NoError(t, err)
+	require.NotEqual(t, cid.Undef, gotPutAdCid1)
+
+	_, err = subject.NotifyPut(ctx, defaultProvider, wantContextID, metadata.New(metadata.Bitswap{}))
+	require.Error(t, err, provider.ErrAlreadyAdvertised)
+
+	p := testutil.NewID(t)
+	_, err = subject.NotifyPut(ctx, peer.AddrInfo{ID: p, Addrs: nil}, wantContextID, metadata.New(metadata.Bitswap{}))
+	require.NoError(t, err, provider.ErrAlreadyAdvertised)
+}
+
+func TestEngine_ShouldHaveSameChunksInChunkerForSameCIDs(t *testing.T) {
+	ctx := contextWithTimeout(t)
+	rng := rand.New(rand.NewSource(1413))
+
+	mhs := testutil.RandomMultihashes(t, rng, 42)
+
+	provider1id := testutil.NewID(t)
+	provider1Addrs, _ := multiaddr.NewMultiaddr("/ip4/0.0.0.0/tcp/1234/http")
+	provider2id := testutil.NewID(t)
+	provider2Addrs, _ := multiaddr.NewMultiaddr("/ip4/0.0.0.0/tcp/4321/http")
+
+	subject, err := engine.New()
+	require.NoError(t, err)
+	err = subject.Start(ctx)
+	require.NoError(t, err)
+	defer subject.Shutdown()
+
+	wantContextID := []byte("fish")
+	subject.RegisterMultihashLister(func(ctx context.Context, p peer.ID, contextID []byte) (provider.MultihashIterator, error) {
+		if string(contextID) == string(wantContextID) {
+			return provider.SliceMultihashIterator(mhs), nil
+		}
+		return nil, errors.New("not found")
+	})
+
+	gotPutAdCid1, err := subject.NotifyPut(ctx, peer.AddrInfo{ID: provider1id, Addrs: []multiaddr.Multiaddr{provider1Addrs}}, wantContextID, metadata.New(metadata.Bitswap{}))
+	require.NoError(t, err)
+	require.NotEqual(t, cid.Undef, gotPutAdCid1)
+	require.Equal(t, 1, subject.Chunker().Len())
+
+	gotPutAdCid2, err := subject.NotifyPut(ctx, peer.AddrInfo{ID: provider2id, Addrs: []multiaddr.Multiaddr{provider2Addrs}}, wantContextID, metadata.New(metadata.Bitswap{}))
+	require.NoError(t, err)
+	require.NotEqual(t, cid.Undef, gotPutAdCid2)
+	require.Equal(t, 1, subject.Chunker().Len())
+	ad1, _ := subject.GetAdv(ctx, gotPutAdCid1)
+	ad2, _ := subject.GetAdv(ctx, gotPutAdCid2)
+
+	require.Equal(t, ad1.Entries, ad2.Entries)
+}
+
+func TestEngine_DatastoreBackwardsCompatibilityTest(t *testing.T) {
+	tempDir, err := ioutil.TempDir("", "DatastoreBackwardCompatibilityTest*")
+	if err != nil {
+		t.Fail()
+	}
+	defer os.RemoveAll(tempDir)
+	// copying testdata into the test dir
+	err = testutil.CopyDir(filepath.Join(testutil.ThisDir(t), "../testdata/datastore.ds"), tempDir)
```

Comment:
I am wondering if we can get away with having a single file that captures the ds. That would be nice, since we won't have to copy the whole directory.

Reply:
two files are required unfortunately